### ISSUE TITLE
WSL env variables are not available at time of docker compose up

### ISSUE THREAD

**Issue by Nvveen** [microsoft/vscode-remote-release#6970]
- VSCode Version: 1.69.2
- Local OS Version: Windows 11 22000.795
- Remote OS Version:
- Remote Extension/Connection Type: Containers/WSL
- Logs:

```
[2022-07-20T11:24:43.535Z] Stop (73 ms): Run: docker compose -f /home/user/work/front/.devcontainer/docker-compose.yml --profile * config
[2022-07-20T11:24:43.535Z] 
[2022-07-20T11:24:43.535Z] time="2022-07-20T13:24:43+02:00" level=warning msg="The \"SSH_AUTH_SOCK\" variable is not set. Defaulting to a blank string."
1 error(s) decoding:

* error decoding 'Volumes[5]': invalid spec: :/ssh-agent: empty section between colons

[2022-07-20T11:24:43.535Z] Exit code 15
[2022-07-20T11:24:43.536Z] Error: Command failed: docker compose -f /home/user/work/front/.devcontainer/docker-compose.yml --profile * config
[2022-07-20T11:24:43.536Z]     at Fa (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:230:907)
[2022-07-20T11:24:43.536Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2022-07-20T11:24:43.536Z]     at async kR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:204:1353)
[2022-07-20T11:24:43.536Z]     at async DR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:188:2228)
[2022-07-20T11:24:43.536Z]     at async QR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:2177)
[2022-07-20T11:24:43.536Z]     at async cb (/home/user/.vscode-remote-containers/dist/dev-containers-cli1.241.3/dist/spec-node/devContainersSpecCLI.js:245:3110)
[2022-07-20T11:24:43.536Z]     at async i_ (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:13194)
[2022-07-20T11:24:43.536Z]     at async n_ (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:12950)
[2022-07-20T11:24:43.538Z] Stop (553 ms): Run in Host: /home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node /home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js up --docker-compose-path docker compose --workspace-folder /home/user/work/front --workspace-mount-consistency cached --id-label devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front --log-level debug --log-format json --config /home/user/work/front/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --remove-existing-container --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-07-20T11:24:43.538Z] Exit code 1
[2022-07-20T11:24:43.541Z] Command failed: /home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node /home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js up --docker-compose-path docker compose --workspace-folder /home/user/work/front --workspace-mount-consistency cached --id-label devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front --log-level debug --log-format json --config /home/user/work/front/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --remove-existing-container --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-07-20T11:24:43.541Z] Exit code 1
[2022-07-20T11:24:53.378Z] Start: Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-20T11:24:53.441Z] Stop (63 ms): Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
```

Steps to Reproduce:

1. Install WSL2 in Windows 11
2. Install VSCode with Remote Containers extension
3. `cd project_dir; code .`
4. (Re)build and (re)open container with the following docker-compose.yml contents:
```
environment:
  - SSH_AUTH_SOCK=/ssh-agent
...
volumes:
  - ${SSH_AUTH_SOCK}:/ssh-agent
```
When I do a manual docker compose up -f .devcontainer/docker-compose.yml --project-name front_devcontainer up, the SSH_AUTH_SOCK variable is set to the containers own ssh-agent socket, although the `/ssh-agent` mountpoint is added and I can `export SSH_AUTH_SOCK=/ssh-agent` to enable ssh forwarding. However, if I rebuild the container inside VSCode, the docker-compose.yml won't have the environment variable from WSL2 available for the 'volume' directive and thus will fail because it sees an empty string.
In short, VSCode won't pass through the WSL Linux environment variables to docker compose to allow in my specific case SSH agent forwarding.

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by chrmarti** [comment]
Could you append the full log?

**Comment by Nvveen** [comment]
```
[2022-07-21T13:42:10.796Z] Remote-Containers 0.241.3 in VS Code 1.69.2 (3b889b090b5ad5793f524b5d1d39fda662b96a2a).
[2022-07-21T13:42:10.796Z] Start: Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:10.862Z] Stop (66 ms): Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:10.871Z] Start: Resolving Remote
[2022-07-21T13:42:10.881Z] Start: Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:10.944Z] Stop (63 ms): Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:10.950Z] Start: Run: wsl -d Ubuntu -e /bin/sh -c cd '/home/user/work/front' && /bin/sh
[2022-07-21T13:42:10.952Z] Start: Run in host: id -un
[2022-07-21T13:42:10.979Z] user
[2022-07-21T13:42:10.979Z] 
[2022-07-21T13:42:10.979Z] Stop (27 ms): Run in host: id -un
[2022-07-21T13:42:10.980Z] Start: Run in host: cat /etc/passwd
[2022-07-21T13:42:10.980Z] Stop (0 ms): Run in host: cat /etc/passwd
[2022-07-21T13:42:10.981Z] Start: Run in host: echo ~
[2022-07-21T13:42:10.982Z] /home/user
[2022-07-21T13:42:10.982Z] 
[2022-07-21T13:42:10.982Z] Stop (1 ms): Run in host: echo ~
[2022-07-21T13:42:10.982Z] Start: Run in host: test -x '/home/user/.vscode-remote-containers/bin/4b889b090b5ad5793f524b5d1d39fda662b96a2a/node'
[2022-07-21T13:42:10.983Z] 
[2022-07-21T13:42:10.983Z] 
[2022-07-21T13:42:10.983Z] Stop (1 ms): Run in host: test -x '/home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node'
[2022-07-21T13:42:10.983Z] Start: Run in host: test -f '/home/user/.vscode-remote-containers/dist/vscode-remote-containers-server-0.241.3.js'
[2022-07-21T13:42:10.984Z] 
[2022-07-21T13:42:10.984Z] 
[2022-07-21T13:42:10.984Z] Stop (1 ms): Run in host: test -f '/home/user/.vscode-remote-containers/dist/vscode-remote-containers-server-0.241.3.js'
[2022-07-21T13:42:10.986Z] userEnvProbe: loginInteractiveShell (default)
[2022-07-21T13:42:10.986Z] userEnvProbe shell: /usr/bin/zsh
[2022-07-21T13:42:11.188Z] userEnvProbe: falling back to printenv
[2022-07-21T13:42:11.188Z] userEnvProbe: loginInteractiveShell (default)
[2022-07-21T13:42:11.188Z] userEnvProbe shell: /usr/bin/zsh
[2022-07-21T13:42:11.318Z] Setting up container for folder or workspace: /home/user/work/front
[2022-07-21T13:42:11.325Z] Start: Check Docker is running
[2022-07-21T13:42:11.325Z] Start: Run in Host: docker version --format {{.Server.APIVersion}}
[2022-07-21T13:42:11.378Z] Stop (53 ms): Run in Host: docker version --format {{.Server.APIVersion}}
[2022-07-21T13:42:11.379Z] Server API version: 1.41
[2022-07-21T13:42:11.379Z] Stop (54 ms): Check Docker is running
[2022-07-21T13:42:11.379Z] Start: Run in Host: docker volume ls -q
[2022-07-21T13:42:11.424Z] Stop (45 ms): Run in Host: docker volume ls -q
[2022-07-21T13:42:11.429Z] Start: Run in Host: docker ps -q -a --filter label=vsch.local.folder=\\wsl.localhost\Ubuntu\home\user\work\front --filter label=vsch.quality=stable
[2022-07-21T13:42:11.478Z] Stop (49 ms): Run in Host: docker ps -q -a --filter label=vsch.local.folder=\\wsl.localhost\Ubuntu\home\user\work\front --filter label=vsch.quality=stable
[2022-07-21T13:42:11.479Z] Start: Run in Host: docker ps -q -a --filter label=devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:11.529Z] Stop (50 ms): Run in Host: docker ps -q -a --filter label=devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:11.532Z] Start: Run in Host: /home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node /home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js up --docker-compose-path docker compose --workspace-folder /home/user/work/front --workspace-mount-consistency cached --id-label devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front --log-level debug --log-format json --config /home/user/work/front/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-07-21T13:42:11.594Z] remote-containers 0.241.3.
[2022-07-21T13:42:11.594Z] Start: Run: docker buildx version
[2022-07-21T13:42:11.670Z] Stop (76 ms): Run: docker buildx version
[2022-07-21T13:42:11.671Z] Start: Resolving Remote
[2022-07-21T13:42:11.673Z] Start: Run: docker compose version --short
[2022-07-21T13:42:11.675Z] Stop (2 ms): Run: docker compose version --short
[2022-07-21T13:42:11.675Z] Start: Run: docker compose version --short
[2022-07-21T13:42:11.677Z] Stop (4 ms): Run: docker compose version --short
[2022-07-21T13:42:11.743Z] Stop (68 ms): Run: docker compose version --short
[2022-07-21T13:42:11.743Z] Start: Run: docker ps -q -a --filter label=com.docker.compose.project=front_devcontainer --filter label=com.docker.compose.service=front
[2022-07-21T13:42:11.810Z] Stop (67 ms): Run: docker ps -q -a --filter label=com.docker.compose.project=front_devcontainer --filter label=com.docker.compose.service=front
[2022-07-21T13:42:11.810Z] Start: Run: docker inspect --type container fe8cd360033d
[2022-07-21T13:42:11.868Z] Stop (58 ms): Run: docker inspect --type container fe8cd360033d
[2022-07-21T13:42:11.868Z] Start: Run: docker compose -f /home/user/work/front/.devcontainer/docker-compose.yml --profile * config
[2022-07-21T13:42:11.940Z] Stop (72 ms): Run: docker compose -f /home/user/work/front/.devcontainer/docker-compose.yml --profile * config
[2022-07-21T13:42:11.940Z] 
[2022-07-21T13:42:11.940Z] time="2022-07-21T15:42:11+02:00" level=warning msg="The \"SSH_AUTH_SOCK\" variable is not set. Defaulting to a blank string."
1 error(s) decoding:

* error decoding 'Volumes[6]': invalid spec: :/ssh-agent: empty section between colons

[2022-07-21T13:42:11.940Z] Exit code 15
[2022-07-21T13:42:11.942Z] Error: Command failed: docker compose -f /home/user/work/front/.devcontainer/docker-compose.yml --profile * config
[2022-07-21T13:42:11.942Z]     at Fa (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:230:907)
[2022-07-21T13:42:11.942Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2022-07-21T13:42:11.942Z]     at async kR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:204:1353)
[2022-07-21T13:42:11.942Z]     at async DR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:188:2228)
[2022-07-21T13:42:11.942Z]     at async QR (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:2177)
[2022-07-21T13:42:11.942Z]     at async cb (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:3110)
[2022-07-21T13:42:11.942Z]     at async i_ (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:13194)
[2022-07-21T13:42:11.942Z]     at async n_ (/home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js:245:12950)
[2022-07-21T13:42:11.944Z] Stop (412 ms): Run in Host: /home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node /home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js up --docker-compose-path docker compose --workspace-folder /home/user/work/front --workspace-mount-consistency cached --id-label devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front --log-level debug --log-format json --config /home/user/work/front/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-07-21T13:42:11.944Z] Exit code 1
[2022-07-21T13:42:11.946Z] Command failed: /home/user/.vscode-remote-containers/bin/3b889b090b5ad5793f524b5d1d39fda662b96a2a/node /home/user/.vscode-remote-containers/dist/dev-containers-cli-0.241.3/dist/spec-node/devContainersSpecCLI.js up --docker-compose-path docker compose --workspace-folder /home/user/work/front --workspace-mount-consistency cached --id-label devcontainer.local_folder=\\wsl.localhost\Ubuntu\home\user\work\front --log-level debug --log-format json --config /home/user/work/front/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root true
[2022-07-21T13:42:11.947Z] Exit code 1
[2022-07-21T13:42:15.084Z] Start: Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
[2022-07-21T13:42:15.147Z] Stop (63 ms): Run: wsl -d Ubuntu -e wslpath -u \\wsl.localhost\Ubuntu\home\user\work\front
```
and the docker-compose.yml file:
```
version: '3.1'
services:
  front:
    container_name: front
    build:
      context: .
      dockerfile: Dockerfile
    user: node
    environment:
        - SSH_AUTH_SOCK=/ssh-agent
    volumes:
      - front-node_modules:/workspace/node_modules
      - vscode-extensions:/home/node/.vscode-server/extensions
      - vscode-extensions-insiders:/home/node/.vscode-server-insiders/extensions
      - "${SSH_AUTH_SOCK}:/ssh-agent"
    ports:
      - '2222:22'
      - '4200:4200'
volumes:
  front-node_modules:
    driver: local
  vscode-extensions:
    driver: local
  vscode-extensions-insiders:
    driver: local
```
I wouldn't really mind not being able to use env variables in docker-compose.yml but I need a workaround to use the SSH agent in the container specifically.

**Comment by corwestermaniddink** [comment]
I had a problem that my local windows system environment variable ${HOME} wasn't available in WSL2. I solved this with running 'setx WSLENV HOME/up' (cmd prompt runas admin) and restarted LxssManager service and restart Docker Desktop.

Maby this solution can also work for your challenge.

**Comment by Nvveen** [comment]
I managed to fix the specific problem with the SSH agent not being forwarded by enabling the agent in Windows (as opposed to the ssh-agent that was running in WSL) and VScode picked this up. However, I'm not sure if that's an exception or if that VSCode still doesn't forward env variables.
